nevergrad/optimization/recastlib.py

```diff
@@ -213,33 +212,141 @@ def _post_loss(self, candidate: p.Parameter, loss: float) -> tp.Loss:
-                )
+class Pymoo(base.ConfiguredOptimizer):
+    """Wrapper over Pymoo optimizer implementations, in standard ask and tell format.
+    This is actually an import from Pymoo Optimize.
 
-            def _evaluate(self, X, out, *args, **kwargs):
-                # pylint:disable=unused-argument
-                # pymoo is supplying us with bounded parameters in [-pi/2,pi/2]. Nevergrad wants unbounded reals from us.
-                out["F"] = self.objective_function(np.tan(X))
+    Parameters
+    ----------
+    algorithm: str
+
+        Use "algorithm-name" with following names to access algorithm classes:
+        Single-Objective
+        -"de"
+        -'ga'
+        -"brkga"
+        -"nelder-mead"
+        -"pattern-search"
+        -"cmaes"
+        Multi-Objective
+        -"nsga2"
+        Multi-Objective requiring reference directions, points or lines
+        -"rnsga2"
+        -"nsga3"
+        -"unsga3"
+        -"rnsga3"
+        -"moead"
+        -"ctaea"
 
-        return _PymooProblem(optimizer, objective_function)
+    Note
+    ----
+    These optimizers do not support asking several candidates in a row
+    """
 
+    recast = True
+    no_parallelization = True
 
-class Pymoo(base.ConfiguredOptimizer):
+    # pylint: disable=unused-argument
+    def __init__(self, *, algorithm: str) -> None:
+        super().__init__(_PymooMinimizeBase, locals())
+
+
+class _PymooBatchMinimizeBase(recaster.BatchRecastOptimizer):
+    def __init__(
+        self,
+        parametrization: IntOrParameter,
+        budget: tp.Optional[int] = None,
+        num_workers: int = 1,
+        *,
+        algorithm: str,
+    ) -> None:
+        super().__init__(parametrization, budget=budget, num_workers=num_workers)
+        # configuration
+        self.algorithm = algorithm
+        self._no_hypervolume = True
+
+    def _internal_tell_not_asked(self, candidate: p.Parameter, loss: tp.Loss) -> None:
+        """Called whenever calling "tell" on a candidate that was not "asked".
+        Defaults to the standard tell pipeline.
+        """  # We do not do anything; this just updates the current best.
+
+    def get_optimization_function(self) -> tp.Callable[[tp.Callable[..., tp.Any]], tp.Optional[tp.ArrayLike]]:
+        # create a different sub-instance, so that the current instance is not referenced by the thread
+        # (consequence: do not create a thread at initialization, or we get a thread explosion)
+        subinstance = self.__class__(
+            parametrization=self.parametrization,
+            budget=self.budget,
+            num_workers=self.num_workers,
+            algorithm=self.algorithm,
+        )
+        # set num_objectives in sub-instance for Pymoo to use in problem definition
+        if self.num_objectives > 0:
+            subinstance.num_objectives = self.num_objectives
+        else:
+            raise RuntimeError("num_objectives should have been set.")
```

Comment:
No need of if/else if one of the branch is just an error, I feel it's simpler to read this way, but not really important:
```suggestion
        if self.num_objectives <= 0:
            raise RuntimeError("num_objectives should have been set.")
        subinstance.num_objectives = self.num_objectives
```

Reply:
I agree!